File: studio/repositories/edits_repo.py
"""Edit log persistence."""

import time
import uuid

from studio import db
from studio.models import Edit, to_doc


def append(
    project_id,
    *,
    source,
    user_prompt,
    action_summary,
    change_scope,
    targets=None,
    version_id=None,
):
    row = Edit(
        editId=uuid.uuid4().hex[:12],
        projectId=project_id,
        versionId=version_id,
        ts=time.time(),
        actor="user",
        source=source,
        userPrompt=user_prompt,
        actionSummary=action_summary,
        changeScope=change_scope,
        targets=targets or [],
    )
    with db.session() as s:
        s.add(row)
        s.commit()
        return to_doc(row)


def list_for_project(project_id):
    with db.session() as s:
        rows = (
            s.query(Edit)
            .filter_by(projectId=project_id)
            .order_by(Edit.ts.desc())
            .all()
        )
        return [to_doc(r) for r in rows]

File: src/server/repos/editsRepo.ts
import { prisma } from "../db";
import { toDoc } from "./doc";

export async function append(
  projectId: string,
  data: {
    source: string;
    userPrompt: string;
    actionSummary: string;
    changeScope: string;
    targets?: string[];
    versionId?: string | null;
  },
): Promise<Record<string, unknown>> {
  const row = await prisma().builderEdit.create({
    data: {
      editId: randomId(),
      projectId,
      versionId: data.versionId ?? null,
      ts: Date.now() / 1000,
      actor: "user",
      source: data.source,
      userPrompt: data.userPrompt,
      actionSummary: data.actionSummary,
      changeScope: data.changeScope,
      targets: data.targets ?? [],
    },
  });
  return toDoc(row) as Record<string, unknown>;
}

export async function listForProject(projectId: string): Promise<Record<string, unknown>[]> {
  const rows = await prisma().builderEdit.findMany({
    where: { projectId },
    orderBy: { ts: "desc" },
  });
  return rows.map((row) => toDoc(row) as Record<string, unknown>);
}

function randomId(): string {
  const buf = new Uint8Array(6);
  crypto.getRandomValues(buf);
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join("");
}
